### PR TITLE
Assert that an op_free()'ed op has an op_first if it has OPf_KIDS

### LINKED ISSUE #20764: `op_free()` silently does nothing if `OPf_KIDS` but `op_first == NULL`
Found while debugging a test failure of `feature-class` branch.

If you want to free the former container of an optree during reshaping, it is insufficient simply to save its child and free the parent by doing this:

```c
OP *kid = cUNOPo->op_first;
cUNOPo->op_first = NULL;
op_free(o);
```

What happens here is that the op still has the `OPf_KIDS` flag set, which causes `Perl_op_free()` to want to recurse into the children. It sets up the pointer to be the `->op_first` of the op, then goes around the while loop again. The loop finds that the pointer is now NULL, decides its job is complete, and the whole function returns having done nothing.

While this is technically the caller's fault (because no op should ever have the `OPf_KIDS` flag while its `op_first` is NULL), I feel that `op_free()` ought to do something about it. Just unsure quite what.

### PR DESCRIPTION
**Pull request by leonerd**
Fixes #20764

### CODE REVIEW

**Comment by leonerd** [comment]
Now with docs added

File: op.c [hunk 1]
```diff
@@ -898,6 +898,11 @@ Perl_op_free(pTHX_ OP *o)
         /* free child ops before ourself, (then free ourself "on the
          * way back up") */
 
+        /* Detect the caller being silly
+         *   https://github.com/Perl/perl5/issues/20764
+         */
```
Comment (tonycoz) [inline op.c]: I think the comment could be improved.  Perhaps:

> Ensure the caller maintains the relationship between OPf_KIDS and op_first when restructuring the tree.

**Comment by demerphq** [comment]
Other than @tonycoz's suggestion (which I agree with), this LGTM.

File: op.c (hunk 1 above)
Comment (leonerd) [reply op.c]: Ahyes; message rewritten and force-pushed.

**Review by tonycoz — APPROVED** [review]

**Review by demerphq — APPROVED** [review]

**Pull request merged**